### PR TITLE
IEEE802.15.4 Framer PSDU Offset Fix

### PR DESCRIPTION
**Pull request by tyler-potyondy**
### Pull Request Overview

This pull request fixes the 15.4 frame data offset calculation. The previous frame metadata was 2 bytes (the same as the PSDU offset). With the recent changes increasing the metadata to 3 bytes, the data offset was incorrect.

### Testing Strategy

This pull request was tested using the OpenThread libtock-c app.

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/extra/src/ieee802154/driver.rs [hunk 1]
```diff
@@ -1143,7 +1143,12 @@ impl<'a, M: device::MacDevice<'a>> device::RxClient for RadioDriver<'a, M> {
                         // the data offset, the data length, and the MIC length.
                         rbuf[(offset + USER_FRAME_METADATA_SIZE)..(offset + user_frame_len)]
                             .copy_from_slice(&buf[PSDU_OFFSET..frame_len]);
-                        rbuf[offset].set(data_offset as u8);
+
+                        // data_offset includes the PSDU offset, so we must subtract this value and add the
+                        // length of the metadata to obtain the data offset for the user frame.
```
Comment (alevy) [inline capsules/extra/src/ieee802154/driver.rs:1148]: This comment is really vague and I don't understand it. What does it mean for `data_offset` to "include" the PSDU offset. The PSDU (PHY Service Data Unit?) is meaningless at this level (this is a client of the Mac device, which has already stripped the PSDU information?

I.e. `data_offset` (which is documented as the first byte in the user frame format, but not clearly defined anywhere in this file what it means) probably should semantically point to the offset where the MAC payload starts.

What is the meaning of this new value that is for some reason?

**Review by alevy — DISMISSED** [review]
> With the recent changes increasing the metadata to 3 bytes, the data offset was incorrect.

I think I'm lacking context here. What are these recent changes? What do they change?

I think from some detective work it is maybe due to a minor addition from #3899 which added the MIC length to the metadata (1 byte) passed up to userpace.

I can see how this might require a different calculation somewhere, but I don't see why it would affect the specifically the `data_offset` value.

**Comment by alevy** [comment]
I think my concern is that this might be fixing a bug in an incremental way, but that the original bug may have actually been the result of introducing information that isn't semantically appropriate at the MAC layer. I'm not certain...

**Comment by tyler-potyondy** [comment]
@alevy in response to your question about `data_offset` not being thoroughly documented in this function, `data_offset` is the offset to the payload. I did not more explicitly document this since this is an argument in the `receive` function and defined in the `RXClient` trait:

```
pub trait RxClient {
    /// When a frame is received, this callback is triggered. The client only
    /// receives an immutable borrow of the buffer. Only completely valid,
    /// unsecured frames that have passed the incoming security procedure are
    /// exposed to the client.
    ///
    /// - `buf`: The entire buffer containing the frame, including extra bytes
    /// in front used for the physical layer.
    /// - `header`: A fully-parsed representation of the MAC header, with the
    /// caveat that the auxiliary security header is still included if the frame
    /// was previously secured.
    /// - `data_offset`: Offset of the data payload relative to
    /// `buf`, so that the payload of the frame is contained in
    /// `buf[data_offset..data_offset + data_len]`.
    /// - `data_len`: Length of the data payload
    fn receive<'a>(&self, buf: &'a [u8], header: Header<'a>, data_offset: usize, data_len: usize);
```

Would it help to have a more explicit comment calling this out within the use in the `receive` function?

**Comment by alevy** [comment]
> Would it help to have a more explicit comment calling this out within the use in the receive function?

So, I'm referring to a _different_ `data_offset`. I'm referring to the one described in the comment in this function showing the structure of a "user_data_frame"---its first field is called `data_offset` and is, presumably, the thing being set to `user_frame_data_offset`.

What is _this_ data-offset? Is it the offset to the data payload? If it is, I don't understand how it _could_ be different from the `data_offset` passed in via the RxClient. Is it, for example, because the bytes we're copying into the user buffer is a subset of the full frame, but not _just_ the subset that is the data payload (e.g. it is preceeded by the MIC?

If so, is the offset we care about not just `MIC_LEN` into the frame?

Basically, I'm starting to believe that this is correct, but I do not have sufficient explanation of what the heck the interface to the process is and what the relationship between that and each of the variables to really understand what's going on.

**Comment by alevy** [comment]
(btw, someone else from, e.g., @tock/networking-wg is welcome to come along and override me. I can't approve without understanding this more, but someone else might be able to, and that's fine)

**Comment by bradjc** [comment]
Why is `PSDU_OFFSET` used at all in this closure? I think if you just delete all uses of that and instead use `data_offset` this issue will go away. I also think the code only works because `PSDU_OFFSET` happens to equal `data_offset`. But I don't think there is a guarantee that will always be true.

Also, it seems like the function argument _should_ be called `mac_data_offset` or similar.

**Comment by tyler-potyondy** [comment]
The terminology for the 15.4 driver and framer have always been confusing to me as well. Thinking more about this, I think all mentions to PSDU offset should be reserved for the radio layer as they just create confusion to buffers and encoding at the framer level.

I'm working to overhaul the framer now and update the terminology to be more clear. I'll push those changes to this branch / PR and update the PR title (unless others would rather I make a separate PR).

**Comment by tyler-potyondy** [comment]
@bradjc PSDU offset is used in the driver receive closure since the `buf` is of the form:

| 2 bytes for radio | PACKET | 

The current design unfortunately passes this up through the entire stack (rather than only passing the packet as the slice). This means that the `data_offset` is the PSDU_OFFSET + HEADER_LEN

**Comment by tyler-potyondy** [comment]
> What is this data-offset? Is it the offset to the data payload? If it is, I don't understand how it could be different from the data_offset passed in via the RxClient. Is it, for example, because the bytes we're copying into the user buffer is a subset of the full frame, but not just the subset that is the data payload (e.g. it is preceeded by the MIC?
>
> If so, is the offset we care about not just MIC_LEN into the frame?

We copy the "full frame" from the perspective of the standard 15.4 frame (i.e. header, payload, mic). We do not copy the 2 bytes that are an artifact of the radio implementation (PSDU offset). I am going to update the 3 metadata elements to now be | header_len | payload_len | mic_len | as this in actuality is the data we need in userland (the current terminology, as evidenced here, creates confusion).

As currently used, the data_offset is the offset to the data payload. This issue arises because of the form of `buf` passed to the driver `receive`

| 2 bytes radio artifact | 15.4 Header | 15.4 Payload | 15.4 MIC | (`buf` passed to `receive`)

We take this and want to form a buffer of the user_frame_form (i.e. possessing the proper metadata so we are able to know what is the header, payload, mic).

We copy into the allowed user buffer

| data_offset | payload_len | mic_len | 15.4 Header | 15.4 Payload | 15.4 MIC | (this is the current implementation. As mentioned before, this will be updated such that data_offset is header_len)

Hopefully this helps provide some more context. (just to clarify, PSDU offset will be removed and this should all be simplified once I complete the changes)

**Comment by tyler-potyondy** [comment]
I was able to remove using `PSDU_OFFSET` in the 15.4 driver receive. However, I was unable to fully refactor `framer.rs` to remove the `PSDU_OFFSET`.

In the case of transmitting, `framer.rs` no longer uses the `PSDU_OFFSET`. Now, the `PSDU_OFFSET` is added in `mac.rs` just prior to the buffer being passed to the radio driver. The receive case was somewhat more complicated unfortunately and would have required an overhaul of the radio drivers, radio HIL, and framer. The current radio HIL design calls the `client.receive` function and expects the client to `set_receive_buffer` with the passed buffer prior to returning. This prevents passing a slice of the radio as this slice will be used to with `set_receive_buffer` (shrinking the buffer each call). 

 In my opinion, any usage/reference to `PSDU_OFFSET` outside direct interactions with the radio object muddles the radio abstraction. I think this should be fixed in the looming 15.4 framer rewrite. For now, it seems acceptable to punt these changes as this is an improvement.

**Review by ppannuto — APPROVED** [review]
This looks like a reasonable fix. I agree that it's not worth sinking too much time into a framer implementation that's due to be replaced (and likely to actually be replaced in a reasonable timeframe).

File: capsules/extra/src/ieee802154/mac.rs [hunk 2]
```diff
@@ -146,6 +146,19 @@ impl<'a, R: radio::Radio<'a>> Mac<'a> for AwakeMac<'a, R> {
         full_mac_frame: &'static mut [u8],
         frame_len: usize,
     ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        // We must add the PSDU_OFFSET required for the radio
```
Comment (brghena) [inline capsules/extra/src/ieee802154/mac.rs:149]: To check that I understand this PR:

1) You have changes in `driver.rs` that properly account for the three-byte header when receiving packets.

2) You have removed the two-byte PDSU header from received packets. This changed `framer.rs` in the `client.receive()` calls.

2) Unrelated to that first two items, you have changed transmitted buffers to only be the data, not a two-byte header and then data. That required changes to `framer.rs` and `mac.rs` but apparently no other files.

Is that right? I think it seems in agreement with the commit history.

**Review by brghena — COMMENTED** [review]
Okay, one suggested change and one check to make sure I understand. But assuming I do, this seems good to go.

I agree that there's some weirdness here about buffers and headers. I think this particular change is fine though, and we don't have to fix every problem just because we're changing something else in the neighborhood. The name changes already implemented to the fields seem to have done enough work.

Overall, the confusions going on here are actually a great example of why the Network WG is developing a new "PacketBuffer" data structure that allows separating these concerns.

**Review by brghena — APPROVED** [review]

**Comment by brghena** [comment]
@alevy Are you okay with merging this now? (either because you agree with it or because you're satisfied that it's been looked at by the proper people)

**Comment by alevy** [comment]
I am
